### ISSUE TITLE
ErrorLog and suppressed errors lead to error in csc.dll

### ISSUE THREAD

**Issue by astolpmann** [dotnet/roslyn#24835]
**Version Used**: 
```
dotnet --info
.NET Command Line Tools (2.1.2)

Product Information:
 Version:            2.1.2
 Commit SHA-1 hash:  5695315371

Runtime Environment:
 OS Name:     Windows
 OS Version:  6.3.9600
 OS Platform: Windows
 RID:         win81-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.1.2\

Microsoft .NET Core Shared Framework Host

  Version  : 2.0.3
  Build    : a9190d4a75f4a982ae4b4fa8d1a24526566c69df
```

**Steps to Reproduce**:

1. ``PS E:\Projekte\!Test\ErrorLogTest> dotnet new console``
2. ``PS E:\Projekte\!Test\ErrorLogTest> dotnet add package Microsoft.CodeAnalysis.FxCopAnalyzers``

3. Disable CA1801 using ``#pragma warning disable CA1801``:
```
using System;
namespace ErrorLogTest
{
    class Program
    {
        #pragma warning disable CA1801
        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
        } }}
```
4. Build the project treating warnings as errors and log analysis report to report.json
``dotnet build /p:TreatWarningsAsErrors=true /p:ErrorLog=report.json``

**Expected Behavior**:
An assembly is build and the CA1801 is logged in report.json

**Actual Behavior**:
```
Build FAILED.
C:\Program Files\dotnet\sdk\2.1.2\Roslyn\Microsoft.CSharp.Core.targets(84,5): error MSB6006: "csc.dll" exited with code 1. [E:\Projekte\!Test\ErrorLogTest\ErrorLogTest.csproj]
    0 Warning(s)
    1 Error(s)
```
The report.json is created correctly regardless and contains the CA1801 error as suppressedInSource.

**Additional Information**:
* This happens with every CA and SA rule.
* Not using ``/p:ErrorLog``, ``/p:TreatWarningsAsErrors=true``, or not suppressing the CA1801 all lead to everything working as expected.

**Comment by jcouv** [comment]
@astolpmann I wasn't able to repro.
Would you mind retrying from your side, maybe with a newer version of `dotnet` to see if you still get the problem?

When running the `build` command, I'm getting:
```
~/repro# /roslyn/Binaries/Tools/dotnet/dotnet build /p:TreatWarningsAsErrors=true /p:ErrorLog=report.json
Microsoft (R) Build Engine version 15.5.178.35674 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restoring packages for /root/repro/repro.csproj...
  Restore completed in 161.42 ms for /root/repro/repro.csproj.
CSC : error CS8032: An instance of analyzer Microsoft.NetCore.Analyzers.Runtime.TestForNaNCorrectlyAnalyzer cannot be created from /root/.nuget/packages/microsoft.netcore.analyzers/2.6.0/analyzers/dotnet/cs/Microsoft.NetCore.Analyzers.dll : Exception has been thrown by the target of an invocation.. [/root/repro/repro.csproj]

Build FAILED.

CSC : error CS8032: An instance of analyzer Microsoft.NetCore.Analyzers.Runtime.TestForNaNCorrectlyAnalyzer cannot be created from /root/.nuget/packages/microsoft.netcore.analyzers/2.6.0/analyzers/dotnet/cs/Microsoft.NetCore.Analyzers.dll : Exception has been thrown by the target of an invocation.. [/root/repro/repro.csproj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:02.13
```

Some theories:
- maybe you were affected by a bug where `dotnet` only reported `csc.dll" exited with code 1` but didn't print the compiler diagnostics. I think that bug was fixed recently.
- maybe I'm using a different version of the package or its dependencies?
- maybe there is something else wrong with my setup.

I'm using 
```
~/repro# /roslyn/Binaries/Tools/dotnet/dotnet --info
.NET Command Line Tools (2.2.0-preview1-007622)

Product Information:
 Version:            2.2.0-preview1-007622
 Commit SHA-1 hash:  045e2b7bf7

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  16.04
 OS Platform: Linux
 RID:         ubuntu.16.04-x64
 Base Path:   /roslyn/Binaries/Tools/dotnet/sdk/2.2.0-preview1-007622/

Microsoft .NET Core Shared Framework Host

  Version  : 2.1.0-preview1-25929-02
  Build    : 328fbf0657c4e3bab243460660583c429eaae001
```

**Comment by astolpmann** [comment]
I would like to do that but can't find the 2.2.0-preview1-007622 (or any 2.2.0) version anywhere. If you can provide me a link i will try it.

**Comment by jcouv** [comment]
I'm getting the same behavior on windows as I did on linux (ie. didn't repro).

```
>dotnet build /p:TreatWarningsAsErrors=true /p:ErrorLog=report.json
Microsoft (R) Build Engine version 15.5.178.35674 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restoring packages for D:\issues\cli\cli.csproj...
  Generating MSBuild file D:\issues\cli\obj\cli.csproj.nuget.g.props.
  Restore completed in 183.74 ms for D:\issues\cli\cli.csproj.
CSC : error CS8032: An instance of analyzer Microsoft.NetCore.Analyzers.Runtime.TestForNaNCorrectlyAnalyzer cannot be created from C:\Users\jcouv\.nuget\packages\microsoft.netcore.analyzers\2.6.0\analyzers\dotnet\cs\Microsoft.NetCore.Analyzers.dll : Exception has been thrown by the target of an invocation.. [D:\issues\cli\cli.csproj]

Build FAILED.

CSC : error CS8032: An instance of analyzer Microsoft.NetCore.Analyzers.Runtime.TestForNaNCorrectlyAnalyzer cannot be created from C:\Users\jcouv\.nuget\packages\microsoft.netcore.analyzers\2.6.0\analyzers\dotnet\cs\Microsoft.NetCore.Analyzers.dll : Exception has been thrown by the target of an invocation.. [D:\issues\cli\cli.csproj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:02.98
```
```
>dotnet --info
.NET Command Line Tools (2.2.0-preview1-007622)

Product Information:
 Version:            2.2.0-preview1-007622
 Commit SHA-1 hash:  045e2b7bf7

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.15063
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.2.0-preview1-007622\

Microsoft .NET Core Shared Framework Host

  Version  : 2.1.0-preview1-25929-02
  Build    : 328fbf0657c4e3bab243460660583c429eaae001
```

To install a 2.1 preview version of CLI, see installers [here](https://github.com/dotnet/cli#installers-and-binaries).

**Comment by astolpmann** [comment]
Im now on another machine with the new version (2.1.300-preview2-008210). Same result. 

```
> dotnet --info
.NET Command Line Tools (2.1.300-preview2-008210)

Product Information:
 Version:            2.1.300-preview2-008210
 Commit SHA-1 hash:  f6065dcc62

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.16299
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.1.300-preview2-008210\

Microsoft .NET Core Shared Framework Host

  Version  : 2.1.0-preview2-26131-06
  Build    : b13a0d5c331f374afd35ded57b9a4b4ab128864c
```

```
> dotnet build /p:TreatWarningsAsErrors=true /p:ErrorLog=report.json
Microsoft (R) Build Engine version 15.6.81.24225 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restoring packages for C:\Users\Andi\Development\ErrorLogTest\ErrorLogTest.csproj...
  Generating MSBuild file C:\Users\Andi\Development\ErrorLogTest\obj\ErrorLogTest.csproj.nuget.g.props.
  Restore completed in 148,34 ms for C:\Users\Andi\Development\ErrorLogTest\ErrorLogTest.csproj.
C:\Program Files\dotnet\sdk\2.1.300-preview2-008210\Roslyn\Microsoft.CSharp.Core.targets(84,5): error MSB6006: "csc.dll" exited with code 1. [C:\Users\Andi\Development\ErrorLogTest\ErrorLogTest.csproj]

Build FAILED.

C:\Program Files\dotnet\sdk\2.1.300-preview2-008210\Roslyn\Microsoft.CSharp.Core.targets(84,5): error MSB6006: "csc.dll" exited with code 1. [C:\Users\Andi\Development\ErrorLogTest\ErrorLogTest.csproj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:03.14
```

**Comment by jcouv** [comment]
I tried again on a fresh machine and I'm now able to repro! Thanks for the help troubleshooting.
Stay tuned.

**Comment by astolpmann** [comment]
I think it is actually you who isn't using the correct version and having problems with: dotnet/roslyn-analyzers#1331 

As you are using
Microsoft (R) Build Engine version 15.5.178.35674 for .NET Core
vs mine
Microsoft (R) Build Engine version 15.6.81.24225 for .NET Core

**Comment by jcouv** [comment]
On the machine where I can repro, I am using the same version as you. Copied the output below.

But something else is strange: On the machine where I can repro, the problem disappears the second time I run the `build` command. This is consistent with every new project I create, and doing `dotnet clean` brings the repro back...

```
C:\Users\jcouv\source\issues\cli2>dotnet build /p:TreatWarningsAsErrors=true /p:ErrorLog=report.json
Microsoft (R) Build Engine version 15.6.81.24225 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restoring packages for C:\Users\jcouv\source\issues\cli2\cli2.csproj...
  Generating MSBuild file C:\Users\jcouv\source\issues\cli2\obj\cli2.csproj.nuget.g.props.
  Restore completed in 118 ms for C:\Users\jcouv\source\issues\cli2\cli2.csproj.
C:\Program Files\dotnet\sdk\2.1.300-preview2-008210\Roslyn\Microsoft.CSharp.Core.targets(84,5): error MSB6006: "csc.dll" exited with code 1. [C:\Users\jcouv\source\issues\cli2\cli2.csproj]

Build FAILED.

C:\Program Files\dotnet\sdk\2.1.300-preview2-008210\Roslyn\Microsoft.CSharp.Core.targets(84,5): error MSB6006: "csc.dll" exited with code 1. [C:\Users\jcouv\source\issues\cli2\cli2.csproj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:00.91

C:\Users\jcouv\source\issues\cli2>dotnet build /p:TreatWarningsAsErrors=true /p:ErrorLog=report.json
Microsoft (R) Build Engine version 15.6.81.24225 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 25.38 ms for C:\Users\jcouv\source\issues\cli2\cli2.csproj.
  cli2 -> C:\Users\jcouv\source\issues\cli2\bin\Debug\netcoreapp2.1\cli2.dll

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:00.69
```
```
C:\Users\jcouv\source\issues\cli2>dotnet --info
.NET Command Line Tools (2.1.300-preview2-008210)

Product Information:
 Version:            2.1.300-preview2-008210
 Commit SHA-1 hash:  f6065dcc62

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.16299
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.1.300-preview2-008210\

Microsoft .NET Core Shared Framework Host

  Version  : 2.1.0-preview2-26131-06
  Build    : b13a0d5c331f374afd35ded57b9a4b4ab128864c
```

**Comment by astolpmann** [comment]
Yes, this also happens for me. This is probably because the project isn't actually build the second time you call "build" as the first time already created the assembly in the obj directory. So the error happens after the actual build and before copying the files to the bin directory.

**Comment by jcouv** [comment]
I've narrowed down the problem a bit. The error behind `error MSB6006: "csc.dll" exited with code 1.` is `error CS2001: Source file 'C:\Users\jcouv\source\issues\cli3\obj\Debug\netcoreapp2.1\cli3.AssemblyInfo.cs' could not be found.`.

The way I arrive at this is `dotnet clean`, then `dotnet build /p:TreatWarningsAsErrors=true /p:ErrorLog=report.json /bl` (to produce a binary log). From the binary log, I get the command-line for `Csc` task and put it in a `.rsp` file (because it's [too long](https://github.com/dotnet/roslyn/wiki/Troubleshooting-tips#running-the-compiler-with-a-long-command-line) to run on the command-line).
Then I do `dotnet clean` followed by `"C:\Program Files\dotnet\dotnet.exe" "C:\Program Files\dotnet\sdk\2.1.300-preview2-008210\Roslyn\bincore\csc.dll" /noconfig @repro.rsp` to simulate that step of the build.
That produces `error CS2001: Source file 'C:\Users\jcouv\source\issues\cli3\obj\Debug\netcoreapp2.1\cli3.AssemblyInfo.cs' could not be found.`

Based on this analysis, I suspect it's a problem with either MSBuild targets or maybe CLI-provided targets. 

@rainersigwald @livarcocc I think this is not a compiler bug. Which repo should I send this issue to?

Note there are two problems here:
1. the build didn't produce a `cli3.AssemblyInfo.cs` file before invoking the compiler on it
2. the `dotnet build` command fails without displaying the diagnostic from the compiler. I believe this is already tracked by https://github.com/dotnet/roslyn/issues/24736

**Comment by rainersigwald** [comment]
@jcouv The symptoms here don't match #24736 -- the errors don't appear in the log and don't appear when building with `/p:UseSharedCompilation=false`.

`dotnet clean` will delete the `.AssemblyInfo.cs` file, so running the compiler directly after doing a clean isn't a good repro for the problem. I did similar to you with some changes:

* set `MSBUILDPRESERVETOOLTEMPFILES=1` to keep the .rsp file from a failing run
* `dotnet clean`
* `dotnet msbuild  /p:TreatWarningsAsErrors=true /p:ErrorLog=report.json` **`/t:GenerateAssemblyInfo`**
* `"C:\Program Files\dotnet\dotnet.exe" "C:\Program Files\dotnet\sdk\2.1.100-preview-007363\Roslyn\bincore\csc.dll" /noconfig @C:\Users\raines\AppData\Local\Temp\tmpe6b69f3bd6364388b82a9d1dd51101ca.rsp`

```
s:\work\roslyn24835>"C:\Program Files\dotnet\dotnet.exe" "C:\Program Files\dotnet\sdk\2.1.100-preview-007363\Roslyn\bincore\csc.dll" /noconfig @C:\Users\raines\AppData\Local\Temp\tmpe6b69f3bd6364388b82a9d1dd51101ca.rsp

s:\work\roslyn24835>echo %ERRORLEVEL%
1

s:\work\roslyn24835>"C:\Program Files\dotnet\dotnet.exe" "C:\Program Files\dotnet\sdk\2.1.100-preview-007363\Roslyn\bincore\csc.dll" /noconfig @C:\Users\raines\AppData\Local\Temp\tmpe6b69f3bd6364388b82a9d1dd51101ca.rsp

s:\work\roslyn24835>echo %ERRORLEVEL%
1
```

That looks like a reproducible failure in Roslyn to me.

**Comment by jcouv** [comment]
Thanks @rainersigwald for the clarification.
I'm debugging through the compiler and getting the following diagnostics:
![image](https://user-images.githubusercontent.com/12466233/36330572-0892ee32-131f-11e8-9d56-d7b2fc677b8b.png)
The one that is an error (and therefore failing the compilation) is CA1801 which comes from an analyzer.

I'm still investigating why the errors are not getting printed out, even when the compiler is run directly.

**Comment by jcouv** [comment]
I've identified the problem (a compiler issue indeed) and am working on a fix. Thanks for your help and patience.

**Comment by jcouv** [comment]
PR for fix is in review (https://github.com/dotnet/roslyn/pull/24908)

**Comment by jcouv** [comment]
Fix merged for 15.7.
